vm/src/builtins.rs

```diff
@@ -583,71 +584,84 @@ fn builtin_pow(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
     }
 }
 
-pub fn builtin_print(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    trace!("print called with {:?}", args);
-
-    // Handle 'sep' kwarg:
-    let sep_arg = args
-        .get_optional_kwarg("sep")
-        .filter(|obj| !obj.is(&vm.get_none()));
-    if let Some(ref obj) = sep_arg {
-        if !objtype::isinstance(obj, &vm.ctx.str_type()) {
-            return Err(vm.new_type_error(format!(
-                "sep must be None or a string, not {}",
-                objtype::get_type_name(&obj.typ())
-            )));
-        }
-    }
-    let sep_str = sep_arg.as_ref().map(|obj| objstr::borrow_value(obj));
-
-    // Handle 'end' kwarg:
-    let end_arg = args
-        .get_optional_kwarg("end")
-        .filter(|obj| !obj.is(&vm.get_none()));
-    if let Some(ref obj) = end_arg {
-        if !objtype::isinstance(obj, &vm.ctx.str_type()) {
-            return Err(vm.new_type_error(format!(
-                "end must be None or a string, not {}",
-                objtype::get_type_name(&obj.typ())
-            )));
+#[derive(Debug)]
+pub struct PrintOptions {
+    sep: Option<PyStringRef>,
+    end: Option<PyStringRef>,
+    flush: bool,
+}
+
+// In the future, this impl will be generated w/ a derive macro.
+impl FromArgs for PrintOptions {
+    fn from_args<I>(
+        vm: &mut VirtualMachine,
+        args: &mut iter::Peekable<I>,
+    ) -> Result<Self, ArgumentError>
+    where
+        I: Iterator<Item = PyArg>,
+    {
+        let mut sep: Option<PyStringRef> = Some(PyString::from(" ").into_ref(vm));
+        let mut end: Option<PyStringRef> = Some(PyString::from("\n").into_ref(vm));
```

Comment:
Shouldn't these be initialized to `None`?

Reply:
No - [the defaults for these params are `" "` and `"\n"`](https://docs.python.org/3/library/functions.html#print). `None` in this context corresponds to a _python_ `None`, which means not to print a separator / end.